R/method-filterFeatures.R

```diff
@@ -0,0 +1,622 @@
+             msg <- NULL
+             if (length(object@threshold) != 1)
+                 msg <- c("'threshold' has to be a numeric of length 1")
+             msg
+         })
+
+#' @rdname BlankFlag
+#'
+#' @export
+BlankFlag <- function(threshold = 2, blankIndex = integer(),
+                         qcIndex = integer(), na.rm = TRUE) {
+    if (is.logical(blankIndex))
+        blankIndex <- which(blankIndex)
+    if(is.numeric(blankIndex))
+        blankIndex <- as.integer(blankIndex)
+    if (is.logical(qcIndex))
+        qcIndex <- which(qcIndex)
+    if(is.numeric(qcIndex))
+        qcIndex <- as.integer(qcIndex)
+    new("BlankFlag", threshold = threshold, blankIndex = blankIndex,
+        qcIndex = qcIndex, na.rm = na.rm)
+}
+
+#' @rdname BlankFlag
+setMethod("filterFeatures",
+          signature(object = "XcmsResult",
+                    filter = "BlankFlag"),
+          function(object, filter, ...){
+              .check_index_range(filter@blankIndex, length(object), name = "blankIndex")
+              .check_index_range(filter@qcIndex, length(object), name = "qcIndex")
+              x <- featureValues(object, ...)[, filter@qcIndex]
+              y <- featureValues(object, ...)[, filter@blankIndex]
+              vals <- rowBlank(x = x, y = y,
+                               na.rm = filter@na.rm,
+                               threshold = filter@threshold)
+              message(sum(vals, na.rm = TRUE) - nrow(featureValues(object)),
+                            " features were flagged")
+              featureDefinitions(object)$possible_contaminants <- vals
+              ph <- XProcessHistory(param = filter,
+                                    date. = date(),
+                                    type. = .PROCSTEP.FEATURE.FILTERING,
+                                    fileIndex = seq_along(object))
+              object <- addProcessHistory(object, ph)
+              validObject(object)
+              object
+          }
+)
+
+#' @rdname BlankFlag
+setMethod("filterFeatures",
+          signature(object = "SummarizedExperiment",
+                    filter = "BlankFlag"),
+          function(object, filter, assay = 1){
+              .check_index_range(filter@blankIndex, ncol(object),
+                                 name = "blankIndex")
+              .check_index_range(filter@qcIndex, ncol(object), name = "qcIndex")
+              x <- assay(object, assay)[, filter@qcIndex]
+              y <- assay(object, assay)[, filter@blankIndex]
+              vals <- rowBlank(x = x, y = y,
+                               na.rm = filter@na.rm,
+                               threshold = filter@threshold)
+              message(sum(vals, na.rm = TRUE) - length(object),
+                      " features were flagged")
+              rowData(object)$possible_contaminants <- vals
+              object
+          }
+)
+
+#' @noRd
+.check_index_range <- function(x, l, name = "") {
+    if (!all(x %in% seq_len(l)) | length(x) == 0)
```

Comment:
as discussed, I would reverse the order of the tests and use `||`, i.e. `if (length(x) == 0 || !all(x %in% seq_len(l)))`